octree/include/pcl/octree/octree_iterator.h

```diff
@@ -524,6 +523,77 @@ namespace pcl
       };
 
     //////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
+    /** \brief @b Octree iterator class
+     * \note This class implements a forward iterator for traversing octrees in a breadth-first manner.
+     *       It only returns the node at a given depth in the octree, if any.
+     * \ingroup octree
+     * \author Fabien Rozar (fabien.rozar@gmail.com)
+     */
+    template<typename OctreeT>
+      class OctreeBreadthFirstAtDepthIterator : public OctreeBreadthFirstIterator<OctreeT>
+      {
+      public:
+
+        // public typedefs
+        typedef typename OctreeBreadthFirstIterator<OctreeT>::BranchNode BranchNode;
+        typedef typename OctreeBreadthFirstIterator<OctreeT>::LeafNode LeafNode;
+
+        /** \brief Empty constructor.
+          * \param[in] depth_arg Depth level during traversal
+          */
+        explicit
+        OctreeBreadthFirstAtDepthIterator (unsigned int depth_arg = 0);
+
+        /** \brief Constructor.
+        * \param[in] octree_arg Octree to be iterated. Initially the iterator is set to its root node.
+        * \param[in] depth_arg Depth level during traversal
+        */
+        explicit
+        OctreeBreadthFirstAtDepthIterator (OctreeT* octree_arg, unsigned int depth_arg = 0);
+
+        /** \brief Empty deconstructor. */
+        virtual
+        ~OctreeBreadthFirstAtDepthIterator ();
+
+        /** \brief Copy operator.
+        * \param[in] src the iterator to copy into this
+        */
+        inline OctreeBreadthFirstAtDepthIterator&
+        operator = (const OctreeBreadthFirstAtDepthIterator& src)
+        {
+          OctreeBreadthFirstIterator<OctreeT>::operator=(src);
+        }
+
+        /** \brief Reset the iterator to the first node at the depth given initially of the octree
+        */
+        void
+        reset ();
+
+        /** \brief Preincrement operator.
+        * \note step to next octree node
+        */
+        OctreeBreadthFirstAtDepthIterator&
+        operator++ ();
+
+        /** \brief postincrement operator.
+        * \note step to next octree node
+        */
+        inline OctreeBreadthFirstAtDepthIterator
+        operator++ (int)
+        {
+          OctreeBreadthFirstAtDepthIterator _Tmp = *this;
+          ++*this;
+          return (_Tmp);
+        }
+
+      protected:
+        using OctreeBreadthFirstIterator<OctreeT>::FIFO_;
+
+        /** \brief Given level of the node to be iterated */
+        unsigned int depth_;
```

Comment:
Can be `const` to make it explicit that depth can not be changed.

Reply:
As an alternative, consider also what I wrote about the reset method and let's decide what fits best.

Reply:
For the reason I gave in the `operator=` comment, the member `depth_` cannot be const.